lib/inventory/inventory.go

```diff
@@ -400,29 +432,39 @@ func (h *downstreamHandle) Close() error {
 	return nil
 }
 
-func (h *downstreamHandle) SendGoodbye(ctx context.Context) error {
+// SendGoodbye crafts a goodbye message, save it, waits for a working stream and sends it to the auth.
+// If the downstreamHandle were to reconnect later, the h.autoEmitGoodbye routine would re-emit it.
+func (h *downstreamHandle) SendGoodbye(ctx context.Context, deleteResources bool, softReload bool) error {
+	goodbye := &proto.UpstreamInventoryGoodbye{DeleteResources: deleteResources, SoftReload: softReload}
+	h.goodbye.Store(goodbye)
+
+	// Wait for an available stream
 	select {
 	case sender := <-h.Sender():
-		// Only send the goodbye if the other half of the stream
-		// has indicated that it supports cleanup. Otherwise, the
-		// upstream will receive an unknown message and terminate
-		// the stream.
-		capabilities := sender.Hello().Capabilities
-		switch {
-		case capabilities == nil:
-			return nil
-		case !capabilities.AppCleanup:
-			return nil
-		}
-
-		return trace.Wrap(sender.Send(ctx, proto.UpstreamInventoryGoodbye{DeleteResources: true}))
+		return trace.Wrap(h.sendGoodbye(ctx, sender, goodbye))
```

Comment:
This can happen concurrently with the `<-h.Sender()` recv in `autoEmitGoodbye`, which will send two goodbyes in the same stream. I don't think it'll lead to anything especially bad, but it's a bit odd.

Should we allow more than one goodbye in the same inventory stream?

Reply:
Yes, I had to adapt the test a bit because this race happens.

> Should we allow more than one goodbye in the same inventory stream?

I think it would be handy if we were to add more info to the goodbye message. The auth implementation is pretty naïve and keeps track of the last goodbye. This would allow us to supersede the previous goodbye message and give new info about our termination process.

If this is an issue I can make `autoEmitGoodbye` block first on `<-handle.Sender()`, then block on waiting the goodbye. But this will make us unable to call SendGoodbye a second time.